cmd/operator/main.go

```diff
@@ -16,46 +16,123 @@ package main
 	stackdriver "github.com/TV4/logrus-stackdriver-formatter"
 	isatty "github.com/mattn/go-isatty"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	log "github.com/sirupsen/logrus"
 )
 
+type stepFlags []int64
+
+func (steps *stepFlags) Set(step string) error {
+	value, err := strconv.ParseInt(step, 10, 64)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse step")
+	}
+	*steps = append(*steps, value)
+	return nil
+}
+
+func (steps stepFlags) String() string {
+	var value string
+	for _, step := range steps {
+		value += " " + strconv.FormatInt(step, 10)
+	}
+
+	return value
+}
+
+var (
+	flCLI         bool
+	flConfigFile  string
+	flHTTPAddr    string
+	flProject     string
+	flRegion      string
+	flService     string
+	flSteps       stepFlags
+	flStepsString string
+	flInterval    int64
+)
+
+func init() {
+	flag.BoolVar(&flCLI, "cli", false, "run as CLI application to manage rollout in intervals")
+	flag.StringVar(&flHTTPAddr, "http-addr", "", "listen on http portrun on request (e.g. :8080)")
+	flag.StringVar(&flProject, "project", "", "project in which the service is deployed")
+	flag.StringVar(&flRegion, "region", "", "the Cloud Run region where the service is deployed")
+	flag.StringVar(&flService, "service", "", "the service to manage")
+	flag.Var(&flSteps, "step", "a percentage in traffic the candidate should go through")
+	flag.StringVar(&flStepsString, "steps", "", "define steps in one flag separated by commas (e.g. 5,30,60)")
+	flag.Int64Var(&flInterval, "interval", 0, "the time between each rollout step")
+	flag.Parse()
+
+	if !flCLI && flHTTPAddr == "" {
+		log.Fatal("one of -cli or -http-addr must be set")
+	}
+
+	if flCLI && flHTTPAddr != "" {
+		log.Fatal("only one of -cli or -http-addr can be used")
+	}
+
+	// -steps flag has precedence over the list of -step flags.
+	if flStepsString != "" {
+		steps := strings.Split(flStepsString, ",")
+		flSteps = []int64{}
+		for _, step := range steps {
+			value, err := strconv.ParseInt(step, 10, 64)
+			if err != nil {
+				log.Fatalf("invalid step value: %v", err)
+			}
+
+			flSteps = append(flSteps, value)
+		}
+	}
+}
+
 func main() {
-	config := &config.Config{
-		Metadata: &config.Metadata{
-			Project: "gtvo-test",
-			Service: "hello",
-			Region:  "us-east1",
-		},
-		Rollout: &config.Rollout{
-			Steps: []int64{5, 30, 80},
-		},
-	}
-
-	log := log.New()
+	logger := log.New()
 	if !isatty.IsTerminal(os.Stdout.Fd()) {
-		log.Formatter = stackdriver.NewFormatter(
+		logger.Formatter = stackdriver.NewFormatter(
 			stackdriver.WithService("cloud-run-release-operator"),
 		)
 	}
 
-	client, err := run.NewAPIClient(context.Background(), config.Metadata.Region)
-	if err != nil {
-		log.Fatalf("could not initilize Cloud Run client: %v", err)
+	if flCLI {
+		runCLI(logger, flConfigFile)
+	}
+}
+
+func runCLI(logger *logrus.Logger, file string) {
+	cfg := config.WithValues(flProject, flRegion, flService, flSteps, flInterval)
+	if !cfg.IsValid(true) {
```

Comment:
I'm not sure why validation is called so late? Can't we validate it when we read them (e.g. in `init()` or as the first thing in `main()`?